allennlp/common/from_params.py

```diff
@@ -0,0 +1,185 @@
+    If it's a function or method, we're checking the object itself.
+    Otherwise, we raise an error.
+    """
+    if inspect.isclass(obj):
+        signature = inspect.signature(obj.__init__)
+    elif inspect.ismethod(obj) or inspect.isfunction(obj):
+        signature = inspect.signature(obj)
+    else:
+        raise ConfigurationError(f"object {obj} is not callable")
+    return arg in signature.parameters
+
+def remove_optional(annotation):
+    """
+    Optional[X] annotations are actually represented as Union[X, NoneType].
+    For our purposes, the "Optional" part is not interesting, so here we
+    throw it away.
+    """
+    origin = getattr(annotation, '__origin__', None)
+    args = getattr(annotation, '__args__', ())
+    if origin == Union and len(args) == 2 and args[1] == type(None):
+        return args[0]
+    else:
+        return annotation
+
+def create_kwargs(cls: Type[T], params: Params, **extras) -> Dict[str, Any]:
+    """
+    Given some class, a `Params` object, and potentially other keyword arguments,
+    create a dict of keyword args suitable for passing to the class's constructor.
+    """
+    # Get the signature of the constructor.
+    signature = inspect.signature(cls.__init__)
+    kwargs: Dict[str, Any] = {}
+
+    # Iterate over all the constructor parameters and their annotations.
+    for name, param in signature.parameters.items():
+        # Skip "self". You're not *required* to call the first parameter "self",
+        # so in theory this logic is fragile, but if you don't call the self parameter
+        # "self" you kind of deserve what happens.
+        if name == "self":
+            continue
+
+        # Get the annotation and break it into `origin` and `args`
+        annotation = remove_optional(param.annotation)
+        origin = getattr(annotation, '__origin__', None)
+        args = getattr(annotation, '__args__', [])
+        default = param.default
+        optional = default != _NO_DEFAULT
+
+        # Some constructors expect extra non-parameter items, e.g. vocab: Vocabulary.
+        # We check the provided `extras` for these and just use them if they exist.
+        if name in extras:
+            kwargs[name] = extras[name]
+
+        # The next case is when the parameter type is itself constructible from_params.
+        elif hasattr(annotation, 'from_params'):
+            if name in params:
+                # Our params have an entry for this, so we use that.
+                subparams = params.pop(name)
+
+                if takes_arg(annotation.from_params, 'extras'):
+                    # If annotation.params accepts **extras, we need to pass them all along.
+                    subextras = extras
+                else:
+                    # Otherwise, only supply the ones that are actual args; any additional ones
+                    # will cause a TypeError.
+                    subextras = {k: v for k, v in extras.items() if takes_arg(annotation.from_params, k)}
```

Comment:
And when should we not crash in this case?  At first glance it feels like we should just be crashing here if we get extra parameters, but I assume you have a good reason for not doing that, and I'm just missing it.  (I may have figured this out; is it because of the vocab passing?)  It still makes me uncomfortable that we can pass `random='garbage'` to these methods without it crashing somewhere.  Will that happen?  Is there a way to make it actually crash?